openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxFastLinkManager.java

```diff
@@ -224,23 +227,57 @@ private void addUriAsLink(RenderingContext c, Box box, PDPage page, float pageHe
 
 				PDAnnotationLink annot = new PDAnnotationLink();
 				annot.setAction(action);
-				if (!placeAnnotation(transform, linkShape, targetArea, annot))
+				if (!placeAnnotation(transform, linkShape, targetArea, new PDAnnotationLinkContainer(annot)))
 					return;
 
-				addLinkToPage(page, annot, box, target);
+				addLinkToPage(page, new PDAnnotationLinkContainer(annot), box, target);
 			} else {
 			    XRLog.general(Level.WARNING, "Could not find valid target for link. Link href = " + uri);
 			}
 		} else if (isURI(uri)) {
-			PDActionURI uriAct = new PDActionURI();
-			uriAct.setURI(uri);
 
 			Rectangle2D targetArea = checkLinkArea(page, c, box, pageHeight, transform, linkShape);
 			if (targetArea == null) {
 				return;
 			}
-			PDAnnotationLink annot = new PDAnnotationLink();
-			annot.setAction(uriAct);
+
+			PDAnnotationLink annotationLink = new PDAnnotationLink();
+			PDActionURI uriAct = new PDActionURI();
+			uriAct.setURI(uri);
+			annotationLink.setAction(uriAct);
+			PDAnnotationContainer annot = new PDAnnotationLinkContainer(annotationLink);
+
+			if ("true".equals(elem.getAttribute("data-embed-file"))) {
+				byte[] file = _sharedContext.getUserAgentCallback().getBinaryResource(uri);
+				if (file != null) {
+					try {
+						PDComplexFileSpecification fs = new PDComplexFileSpecification();
+						PDEmbeddedFile embeddedFile = new PDEmbeddedFile(_od.getWriter(), new ByteArrayInputStream(file));
+						embeddedFile.setSubtype(elem.getAttribute("data-content-type") != null ? elem.getAttribute("data-content-type") : "application/octet-stream");
+						fs.setEmbeddedFile(embeddedFile);
+						String fileName = Paths.get(uri).getFileName().toString();
```

Comment:
This won't work for custom protocol handlers reliably. I think it may be better to use another attribute, possibly `download`.

Reply:
You are right, it may fail. Maybe `data-file-name` as an attribute is more telling?

Reply:
I agree `download` isn't the best attribute name, but it is the standard html5 attribute for this purpose.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#Attributes

Reply:
I didn't know about the download attribute :D . So I could remove "data-embed-file" and use the `download` attribute with a name.